signalfx/resource_signalfx_aws_integration.go

```diff
@@ -678,78 +757,38 @@ func DoIntegrationAWSDelete(d *schema.ResourceData, meta interface{}) error {
 	// Disable the AWS logs sync and/or CloudWatch metric streams sync if needed
 	needToDisableMetricStreams := int.Enabled && int.MetricStreamsSyncState != "" && int.MetricStreamsSyncState != "DISABLED"
 	needToDisableLogsSync := int.Enabled && int.LogsSyncState != "" && int.LogsSyncState != "DISABLED"
+	var toWaitFor []TargetState
 	if needToDisableMetricStreams || needToDisableLogsSync {
 		if needToDisableMetricStreams {
 			int.MetricStreamsSyncState = "CANCELLING"
+			toWaitFor = append(toWaitFor, TargetState{"metric_streams_sync_state", "disabled"})
 		}
 		if needToDisableLogsSync {
 			int.LogsSyncState = "CANCELLING"
-		}
-		_, err := config.Client.UpdateAWSCloudWatchIntegration(context.TODO(), d.Id(), int)
-		if err != nil {
-			if strings.Contains(err.Error(), "40") {
-				err = fmt.Errorf("%s\nPlease verify you are using an admin token when working with integrations", err.Error())
-			}
-			return err
-		}
-		if needToDisableMetricStreams {
-			if _, err = waitForIntegrationSpecificSyncStateToSettle(d, false, config, int.Id, "use_metric_streams_sync", metricStreamsStateSupplier); err != nil {
-				return err
-			}
-		}
-		if needToDisableLogsSync {
-			if _, err = waitForIntegrationSpecificSyncStateToSettle(d, false, config, int.Id, "enable_logs_sync", logsSyncStateSupplier); err != nil {
-				return err
-			}
+			toWaitFor = append(toWaitFor, TargetState{"logs_sync_state", "disabled"})
 		}
 	}
 
-	return config.Client.DeleteAWSCloudWatchIntegration(context.TODO(), d.Id())
-}
-
-func waitForIntegrationStateToSettle(d *schema.ResourceData, config *signalfxConfig, intId string, syncStateField string,
-	stateSupplier stateSupplierFunc) (*integration.AwsCloudWatchIntegration, error) {
-	return waitForIntegrationSpecificSyncStateToSettle(d, d.Get(syncStateField).(bool), config, intId, syncStateField, stateSupplier)
-}
-
-func waitForIntegrationSpecificSyncStateToSettle(d *schema.ResourceData, syncState bool, config *signalfxConfig, intId string, syncStateField string,
-	stateSupplier stateSupplierFunc) (*integration.AwsCloudWatchIntegration, error) {
-	var pending, target []string
-	var expectedState string
-	if syncState {
-		expectedState = "enabled"
-		pending = []string{"DISABLED", "CANCELLING", "CANCELLATION_FAILED"}
-		target = []string{"ENABLED"}
-	} else {
-		expectedState = "disabled"
-		pending = []string{"ENABLED", "CANCELLING"}
-		target = []string{"", "DISABLED"}
-	}
-
-	stateConf := &resource.StateChangeConf{
-		Pending: pending,
-		Target:  target,
-		Refresh: func() (interface{}, string, error) {
-			int, err := config.Client.GetAWSCloudWatchIntegration(context.TODO(), intId)
-			if err != nil {
-				return 0, "", err
-			}
-			return int, stateSupplier(int), nil
-		},
-		Timeout:    d.Timeout(schema.TimeoutUpdate) - time.Minute,
-		Delay:      10 * time.Second,
-		MinTimeout: 5 * time.Second,
+	// Always disable integration before deleting it. This is crucial in case of usage of Metric Streams
```

Comment:
not sure if this is required ihmo it is enough to have logs and streams disabled